qutip/wigner.py

```diff
@@ -548,83 +548,297 @@ def _wig_laguerre_val(L, x, c):
+
+    Examples
+    --------
+    Initialise the class for a square set of coordinates, with some states we
+    want to investigate.
 
-    xvec : array_like
-        x-coordinates at which to calculate the Husimi-Q function.
+    >>> xs = np.linspace(-2, 2, 101)
+    >>> states = [qutip.rand_dm(10) for _ in [None]*10]
+    >>> qfunc = qutip.QFunc(xs, xs)
 
-    yvec : array_like
-        y-coordinates at which to calculate the Husimi-Q function.
-
-    g : float
-        Scaling factor for `a = 0.5 * g * (x + iy)`, default `g = sqrt(2)`.
-        The value of `g` is related to the value of `hbar` in the commutation
-        relation `[x, y] = 1j * hbar` via `hbar=2/g^2` giving the default
-        value `hbar=1`.
+    Now we can calculate the Husimi-Q function over each of the states more
+    efficiently with:
 
+    >>> husimiq = np.array([qfunc(state) for state in states])
 
-    Returns
+    See Also
     --------
-    Q : array
-        Values representing the Q-function calculated over the specified range
-        [xvec,yvec].
-
+    :obj:`.qfunc` :
+        a single function version, which will involve computing several
+        quantities multiple times in order to use less memory.
     """
-    X, Y = meshgrid(xvec, yvec)
-    amat = 0.5 * g * (X + Y * 1j)
-
-    if not (isoper(state) or isket(state)):
-        raise TypeError('Invalid state operand to qfunc.')
-
-    qmat = zeros(size(amat))
-
-    if isket(state):
-        qmat = _qfunc_pure(state, amat)
-    elif isoper(state):
-        d, v = la.eig(state.full())
-        # d[i]   = eigenvalue i
-        # v[:,i] = eigenvector i
-
-        qmat = zeros(np.shape(amat))
-        for k in arange(0, len(d)):
-            qmat1 = _qfunc_pure(v[:, k], amat)
-            qmat += real(d[k] * qmat1)
+    def __init__(self, xs, ys, g: float = np.sqrt(2), memory: float = 1024):
+        self._g = g
+        self._alpha_matrix = _qfunc_alpha_matrix(xs, ys, g)
+        # 16 bytes per complex, 1024**2 bytes per MB.
+        self._size_mb = self._alpha_matrix.conj.size * 16 / (1024**2)
+        self._memory_mb = memory
+        self._max_size = int(self._memory_mb // self._size_mb)
+        self._current_size = 0
+        self._cache = None
+
+    def _alphas(self, size: int):
+        r"""
+        Retrive the full tensor of (the conjugate of) coherent states over all
+        values of :math:`\alpha`, for states of dimension ``size``.
+        """
+        if self._current_size >= size:
+            return self._cache[:, :, :size]
+        if size > self._max_size:
+            requirement = self._size_mb * size
+            raise MemoryError(
+                f"Refusing to precompute up to {size} basis states."
+                f" This would require {requirement:.2f} MB,"
+                f" but only {self._memory_mb} MB is allowed."
+            )
+        if self._cache is None:
+            self._cache = self._alpha_matrix(self._current_size, size)
```

Comment:
Could we add a comment here saying self._cache is a [n, m, k] sized numpy array of complex numbers where (n, m) represents the phase space grid and k represents the Hilbert space dimensions for the coherent state.